File: hypr-control-center/src/waybar-taskbar copy 2/src/icon_resolver.cpp
/**
 * ═══════════════════════════════════════════════════════════════════════════════
 * Icon Resolver - Implementation
 * ═══════════════════════════════════════════════════════════════════════════════
 */

#include "icon_resolver.hpp"

#include <gtk/gtk.h>
#include <gio/gio.h>
#include <gio/gdesktopappinfo.h>
#include <filesystem>
#include <algorithm>
#include <fstream>
#include <cstdlib>

namespace fs = std::filesystem;

IconResolver::IconResolver() {
    init_nerd_icons();
    init_wm_class_map();
    detect_icon_theme();
}

IconResolver::~IconResolver() {}

void IconResolver::init_nerd_icons() {
    // Browsers
    m_nerd_icons["firefox"] = "";
    m_nerd_icons["chromium"] = "";
    m_nerd_icons["google-chrome"] = "";
    m_nerd_icons["brave-browser"] = "󰖟";
    m_nerd_icons["zen-browser"] = "󰖟";
    
    // Terminals
    m_nerd_icons["kitty"] = "";
    m_nerd_icons["alacritty"] = "";
    m_nerd_icons["foot"] = "";
    m_nerd_icons["wezterm"] = "";
    m_nerd_icons["konsole"] = "";
    
    // Development
    m_nerd_icons["code"] = "󰨞";
    m_nerd_icons["visual-studio-code"] = "󰨞";
    m_nerd_icons["code-oss"] = "󰨞";
    m_nerd_icons["neovim"] = "";
    m_nerd_icons["vim"] = "";
    
    // Files
    m_nerd_icons["nautilus"] = "";
    m_nerd_icons["thunar"] = "";
    m_nerd_icons["dolphin"] = "";
    m_nerd_icons["pcmanfm"] = "";
    m_nerd_icons["nemo"] = "";
    
    // Media
    m_nerd_icons["spotify"] = "";
    m_nerd_icons["vlc"] = "󰕼";
    m_nerd_icons["mpv"] = "";
    
    // Communication
    m_nerd_icons["discord"] = "󰙯";
    m_nerd_icons["slack"] = "󰒱";
    m_nerd_icons["telegram-desktop"] = "";
    m_nerd_icons["signal-desktop"] = "󰭹";
    
    // System
    m_nerd_icons["pavucontrol"] = "󰕾";
    m_nerd_icons["blueman-manager"] = "";
    
    // Graphics
    m_nerd_icons["gimp"] = "";
    m_nerd_icons["inkscape"] = "";
    m_nerd_icons["blender"] = "󰂫";
    
    // Games
    m_nerd_icons["steam"] = "";
    m_nerd_icons["lutris"] = "";
    
    // Default
    m_nerd_icons["default"] = "󰣆";
}

void IconResolver::init_wm_class_map() {
    m_wm_class_map["Firefox"] = "firefox";
    m_wm_class_map["firefox"] = "firefox";
    m_wm_class_map["Google-chrome"] = "google-chrome";
    m_wm_class_map["Brave-browser"] = "brave-browser";
    m_wm_class_map["Code"] = "visual-studio-code";
    m_wm_class_map["code"] = "visual-studio-code";
    m_wm_class_map["code-oss"] = "code-oss";
    m_wm_class_map["kitty"] = "kitty";
    m_wm_class_map["Alacritty"] = "Alacritty";
    m_wm_class_map["org.gnome.Nautilus"] = "org.gnome.Nautilus";
    m_wm_class_map["Thunar"] = "Thunar";
    m_wm_class_map["thunar"] = "Thunar";
    m_wm_class_map["Discord"] = "discord";
    m_wm_class_map["discord"] = "discord";
    m_wm_class_map["Spotify"] = "spotify";
    m_wm_class_map["spotify"] = "spotify";
    m_wm_class_map["Steam"] = "steam";
    m_wm_class_map["steam"] = "steam";
    m_wm_class_map["TelegramDesktop"] = "telegram";
    m_wm_class_map["Gimp-2.10"] = "gimp";
    m_wm_class_map["Zen Browser"] = "zen-browser";
}

void IconResolver::detect_icon_theme() {
    m_current_theme = "Adwaita"; // Default
    
    // Try gsettings
    GSettings* settings = g_settings_new("org.gnome.desktop.interface");
    if (settings) {
        gchar* theme = g_settings_get_string(settings, "icon-theme");
        if (theme) {
            m_current_theme = theme;
            g_free(theme);
        }
        g_object_unref(settings);
    }
}

std::string IconResolver::get_current_theme() {
    return m_current_theme;
}

std::string IconResolver::normalize_app_id(const std::string& app_id) {
    // Check wm_class map first
    auto it = m_wm_class_map.find(app_id);
    if (it != m_wm_class_map.end()) {
        return it->second;
    }
    
    // Lowercase and replace underscores
    std::string normalized = app_id;
    std::transform(normalized.begin(), normalized.end(), normalized.begin(), ::tolower);
    std::replace(normalized.begin(), normalized.end(), '_', '-');
    
    return normalized;
}

std::string IconResolver::get_icon_path(const std::string& app_id, int size) {
    // Check cache
    std::string cache_key = app_id + ":" + std::to_string(size);
    auto it = m_cache.find(cache_key);
    if (it != m_cache.end()) {
        return it->second;
    }
    
    std::string normalized = normalize_app_id(app_id);
    std::string path;
    
    // Search order
    path = find_in_theme(normalized, size);
    if (path.empty()) path = find_in_hicolor(normalized, size);
    if (path.empty()) path = find_in_pixmaps(normalized);
    if (path.empty()) path = find_desktop_icon(app_id);
    
    // Cache result
    m_cache[cache_key] = path;
    
    return path;
}

std::string IconResolver::find_in_theme(const std::string& app_id, int size) {
    std::vector<std::string> theme_paths = {
        std::string(getenv("HOME")) + "/.local/share/icons/" + m_current_theme,
        std::string(getenv("HOME")) + "/.icons/" + m_current_theme,
        "/usr/share/icons/" + m_current_theme,
    };
    
    std::vector<std::string> size_dirs = {
        std::to_string(size) + "x" + std::to_string(size) + "/apps",
        "scalable/apps",
        "48x48/apps",
        "32x32/apps",
        "24x24/apps",
        "22x22/apps",
    };
    
    std::vector<std::string> extensions = {".svg", ".png", ".xpm"};
    
    // Try variations of app_id
    std::vector<std::string> app_ids = {
        app_id,
        normalize_app_id(app_id),
    };
    
    // Add org.gnome variant
    std::string capitalized = app_id;
    if (!capitalized.empty()) {
        capitalized[0] = std::toupper(capitalized[0]);
    }
    app_ids.push_back("org.gnome." + capitalized);
    
    for (const auto& theme_path : theme_paths) {
        if (!fs::exists(theme_path)) continue;
        
        for (const auto& size_dir : size_dirs) {
            fs::path icon_dir = fs::path(theme_path) / size_dir;
            if (!fs::exists(icon_dir)) continue;
            
            for (const auto& aid : app_ids) {
                for (const auto& ext : extensions) {
                    fs::path icon_path = icon_dir / (aid + ext);
                    if (fs::exists(icon_path)) {
                        return icon_path.string();
                    }
                }
            }
        }
    }
    
    return "";
}

std::string IconResolver::find_in_hicolor(const std::string& app_id, int size) {
    std::vector<std::string> hicolor_paths = {
        "/usr/share/icons/hicolor",
        std::string(getenv("HOME")) + "/.local/share/icons/hicolor",
    };
    
    std::vector<std::string> size_dirs = {
        std::to_string(size) + "x" + std::to_string(size) + "/apps",
        "scalable/apps",
        "48x48/apps",
        "32x32/apps",
    };
    
    std::vector<std::string> extensions = {".svg", ".png", ".xpm"};
    
    for (const auto& hicolor_path : hicolor_paths) {
        if (!fs::exists(hicolor_path)) continue;
        
        for (const auto& size_dir : size_dirs) {
            fs::path icon_dir = fs::path(hicolor_path) / size_dir;
            if (!fs::exists(icon_dir)) continue;
            
            for (const auto& ext : extensions) {
                fs::path icon_path = icon_dir / (app_id + ext);
                if (fs::exists(icon_path)) {
                    return icon_path.string();
                }
            }
        }
    }
    
    return "";
}

std::string IconResolver::find_in_pixmaps(const std::string& app_id) {
    std::vector<std::string> extensions = {".png", ".svg", ".xpm"};
    
    for (const auto& ext : extensions) {
        fs::path icon_path = fs::path("/usr/share/pixmaps") / (app_id + ext);
        if (fs::exists(icon_path)) {
            return icon_path.string();
        }
    }
    
    return "";
}

std::string IconResolver::find_desktop_icon(const std::string& app_id) {
    // Try to find .desktop file and get icon from there
    std::vector<std::string> desktop_paths = {
        std::string(getenv("HOME")) + "/.local/share/applications",
        "/usr/share/applications",
    };
    
    std::vector<std::string> variations = {
        app_id + ".desktop",
        normalize_app_id(app_id) + ".desktop",
    };
    
    for (const auto& desktop_path : desktop_paths) {
        if (!fs::exists(desktop_path)) continue;
        
        for (const auto& variation : variations) {
            fs::path desktop_file = fs::path(desktop_path) / variation;
            if (fs::exists(desktop_file)) {
                // Parse desktop file for Icon=
                GDesktopAppInfo* app_info = g_desktop_app_info_new_from_filename(desktop_file.c_str());
                if (app_info) {
                    const char* icon_name = g_desktop_app_info_get_string(app_info, "Icon");
                    if (icon_name) {
                        std::string icon_str = icon_name;
                        g_object_unref(app_info);
                        
                        // If it's a full path, return it
                        if (icon_str[0] == '/') {
                            return icon_str;
                        }
                        
                        // Otherwise search for it
                        return find_in_theme(icon_str, 24);
                    }
                    g_object_unref(app_info);
                }
            }
        }
    }
    
    return "";
}

std::string IconResolver::get_nerd_icon(const std::string& app_id) {
    std::string normalized = normalize_app_id(app_id);
    
    // Direct match
    auto it = m_nerd_icons.find(normalized);
    if (it != m_nerd_icons.end()) {
        return it->second;
    }
    
    // Partial match
    for (const auto& [key, icon] : m_nerd_icons) {
        if (normalized.find(key) != std::string::npos || key.find(normalized) != std::string::npos) {
            return icon;
        }
    }
    
    return m_nerd_icons["default"];
}

void IconResolver::clear_cache() {
    m_cache.clear();
}

File: .config/hypr-control-center/src/zenpybarv2/core/icon_resolver.py
"""
IconResolver - Application Icon Resolution for ZenPyBar v2
==========================================================

Resolves application icons from multiple sources:
1. Custom icons directory
2. Current GTK icon theme (Papirus, Tela, etc.)
3. Hicolor fallback
4. /usr/share/pixmaps
5. Nerd Font fallback

Uses caching for performance.
"""

import os
from pathlib import Path
from typing import Dict, Optional, List
import threading


class IconResolver:
    """
    Resolves application icons from system icon themes.
    
    SINGLETON - shared across all components for cache efficiency.
    """
    
    # Common icon search paths
    ICON_PATHS = [
        Path.home() / ".config/hypr-control-center/icons",
        Path.home() / ".local/share/icons",
        Path.home() / ".icons",
        Path("/usr/share/icons"),
        Path("/usr/share/pixmaps"),
    ]
    
    # Nerd Font icon fallbacks
    NERD_ICONS = {
        # Browsers
        'firefox': '',
        'firefox-developer-edition': '',
        'chromium': '',
        'google-chrome': '',
        'brave': '󰖟',
        'brave-browser': '󰖟',
        'vivaldi': '󰖟',
        'microsoft-edge': '󰇩',
        'zen-browser': '󰖟',
        'zen': '󰖟',
        
        # Terminals
        'kitty': '',
        'alacritty': '',
        'foot': '',
        'wezterm': '',
        'konsole': '',
        'gnome-terminal': '',
        'xterm': '',
        'terminator': '',
        
        # Development
        'code': '󰨞',
        'code-oss': '󰨞',
        'visual-studio-code': '󰨞',
        'vscodium': '󰨞',
        'sublime-text': '',
        'neovim': '',
        'nvim': '',
        'vim': '',
        'emacs': '',
        'jetbrains-idea': '',
        'jetbrains-pycharm': '',
        'jetbrains-clion': '',
        'jetbrains-webstorm': '',
        'android-studio': '',
        
        # Files & Editors
        'nautilus': '',
        'org.gnome.nautilus': '',
        'thunar': '',
        'dolphin': '',
        'pcmanfm': '',
        'nemo': '',
        'caja': '',
        'gedit': '',
        'kate': '',
        'mousepad': '',
        'geany': '',
        
        # Media
        'spotify': '',
        'vlc': '󰕼',
        'mpv': '',
        'rhythmbox': '󰎆',
        'audacious': '󰎆',
        'clementine': '󰎆',
        'elisa': '󰎆',
        'amberol': '󰎆',
        
        # Communication
        'discord': '󰙯',
        'slack': '󰒱',
        'telegram-desktop': '',
        'telegramdesktop': '',
        'signal-desktop': '󰭹',
        'element': '',
        'teams': '󰊻',
        'zoom': '',
        'skype': '󰒯',
        
        # System
        'gnome-control-center': '',
        'systemsettings': '',
        'pavucontrol': '󰕾',
        'blueman-manager': '',
        'nm-connection-editor': '󰖩',
        'gnome-system-monitor': '',
        
        # Graphics
        'gimp': '',
        'gimp-2.10': '',
        'inkscape': '',
        'blender': '󰂫',
        'krita': '',
        
        # Office
        'libreoffice-writer': '󰈙',
        'libreoffice-calc': '󰧷',
        'libreoffice-impress': '󰐩',
        'libreoffice-draw': '',
        'onlyoffice': '󰈙',
        
        # Games
        'steam': '',
        'lutris': '',
        'heroic': '',
        
        # Utilities
        'obs': '󰑋',
        'obs-studio': '󰑋',
        'com.obsproject.studio': '󰑋',
        'flameshot': '',
        'grim': '',
        'keepassxc': '󰌋',
        'bitwarden': '󰌋',
        '1password': '󰌋',
        
        # Default
        'default': '󰣆',
    }
    
    # Known wm_class to icon name mappings
    WM_CLASS_MAP = {
        'firefox': 'firefox',
        'Firefox Developer Edition': 'firefox-developer-edition',
        'google-chrome': 'google-chrome',
        'Google-chrome': 'google-chrome',
        'chromium': 'chromium',
        'Brave-browser': 'brave-browser',
        'code': 'visual-studio-code',
        'Code': 'visual-studio-code',
        'code-oss': 'code-oss',
        'kitty': 'kitty',
        'Alacritty': 'Alacritty',
        'foot': 'foot',
        'org.gnome.Nautilus': 'org.gnome.Nautilus',
        'thunar': 'Thunar',
        'Thunar': 'Thunar',
        'discord': 'discord',
        'Discord': 'discord',
        'Spotify': 'spotify',
        'spotify': 'spotify',
        'vlc': 'vlc',
        'mpv': 'mpv',
        'Gimp-2.10': 'gimp',
        'steam': 'steam',
        'Steam': 'steam',
        'obs': 'com.obsproject.Studio',
        'telegram-desktop': 'telegram',
        'TelegramDesktop': 'telegram',
        'zen-browser': 'zen-browser',
        'Zen Browser': 'zen-browser',
    }
    
    _instance = None
    _lock = threading.Lock()
    
    def __new__(cls, *args, **kwargs):
        """Singleton pattern"""
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if self._initialized:
            return
        
        self._cache: Dict[str, Optional[str]] = {}
        self._current_theme: Optional[str] = None
        
        # Detect current icon theme
        self._detect_icon_theme()
        
        self._initialized = True
        print(f"[IconResolver] ✅ Initialized (theme: {self._current_theme})")
    
    def _detect_icon_theme(self) -> None:
        """Detect current GTK icon theme"""
        import subprocess
        
        # Try gsettings
        try:
            result = subprocess.run(
                ['gsettings', 'get', 'org.gnome.desktop.interface', 'icon-theme'],
                capture_output=True, text=True, timeout=2
            )
            if result.returncode == 0:
                self._current_theme = result.stdout.strip().strip("'\"")
                return
        except Exception:
            pass
        
        # Try GTK settings file
        gtk_settings = Path.home() / ".config/gtk-4.0/settings.ini"
        if gtk_settings.exists():
            try:
                content = gtk_settings.read_text()
                for line in content.split('\n'):
                    if line.startswith('gtk-icon-theme-name'):
                        self._current_theme = line.split('=')[1].strip()
                        return
            except Exception:
                pass
        
        # Fallback
        self._current_theme = "Adwaita"
    
    # ═══════════════════════════════════════════════════════════════════════
    # PUBLIC API
    # ═══════════════════════════════════════════════════════════════════════
    
    def get_icon_path(self, app_id: str, size: int = 24) -> Optional[str]:
        """
        Get icon path for an application.
        
        Args:
            app_id: Application ID or wm_class
            size: Desired icon size
            
        Returns:
            Path to icon file or None
        """
        cache_key = f"{app_id}:{size}"
        
        if cache_key in self._cache:
            return self._cache[cache_key]
        
        # Normalize app_id
        normalized = self._normalize_app_id(app_id)
        
        # Search in order
        path = None
        
        # 1. Custom icons
        path = self._find_in_custom(normalized, size)
        
        # 2. Current theme
        if not path:
            path = self._find_in_theme(normalized, size)
        
        # 3. Hicolor fallback
        if not path:
            path = self._find_in_hicolor(normalized, size)
        
        # 4. Pixmaps
        if not path:
            path = self._find_in_pixmaps(normalized)
        
        # Cache result
        self._cache[cache_key] = path
        
        return path
    
    def get_nerd_icon(self, app_id: str) -> str:
        """Get Nerd Font icon fallback"""
        normalized = app_id.lower().replace(' ', '-')
        
        # Direct match
        if normalized in self.NERD_ICONS:
            return self.NERD_ICONS[normalized]
        
        # Partial match
        for key, icon in self.NERD_ICONS.items():
            if key in normalized or normalized in key:
                return icon
        
        return self.NERD_ICONS['default']
    
    def clear_cache(self) -> None:
        """Clear icon cache"""
        self._cache.clear()
    
    # ═══════════════════════════════════════════════════════════════════════
    # INTERNAL SEARCH METHODS
    # ═══════════════════════════════════════════════════════════════════════
    
    def _normalize_app_id(self, app_id: str) -> str:
        """Normalize app ID for searching"""
        if app_id in self.WM_CLASS_MAP:
            return self.WM_CLASS_MAP[app_id]
        
        return app_id.lower().replace(' ', '-').replace('_', '-')
    
    def _find_in_custom(self, app_id: str, size: int) -> Optional[str]:
        """Find in custom icons directory"""
        custom_dir = Path.home() / ".config/hypr-control-center/icons"
        if not custom_dir.exists():
            return None
        
        for ext in ['.png', '.svg', '.xpm']:
            icon_path = custom_dir / f"{app_id}{ext}"
            if icon_path.exists():
                return str(icon_path)
        
        return None
    
    def _find_in_theme(self, app_id: str, size: int) -> Optional[str]:
        """Find in current icon theme"""
        if not self._current_theme:
            return None
        
        theme_paths = [
            Path.home() / ".local/share/icons" / self._current_theme,
            Path.home() / ".icons" / self._current_theme,
            Path("/usr/share/icons") / self._current_theme,
        ]
        
        size_dirs = [
            f"{size}x{size}/apps",
            "scalable/apps",
            "48x48/apps",
            "32x32/apps",
            "24x24/apps",
            "22x22/apps",
            "16x16/apps",
        ]
        
        # Also try variations of app_id
        app_ids_to_try = [
            app_id,
            app_id.lower(),
            app_id.replace('-', '.'),
            f"org.gnome.{app_id.capitalize()}",
        ]
        
        for theme_path in theme_paths:
            if not theme_path.exists():
                continue
            
            for size_dir in size_dirs:
                icon_dir = theme_path / size_dir
                if not icon_dir.exists():
                    continue
                
                for aid in app_ids_to_try:
                    for ext in ['.svg', '.png', '.xpm']:
                        icon_path = icon_dir / f"{aid}{ext}"
                        if icon_path.exists():
                            return str(icon_path)
        
        return None
    
    def _find_in_hicolor(self, app_id: str, size: int) -> Optional[str]:
        """Find in hicolor theme"""
        hicolor_paths = [
            Path("/usr/share/icons/hicolor"),
            Path.home() / ".local/share/icons/hicolor",
        ]
        
        size_dirs = [
            f"{size}x{size}/apps",
            "scalable/apps",
            "48x48/apps",
            "32x32/apps",
            "24x24/apps",
        ]
        
        for hicolor_path in hicolor_paths:
            if not hicolor_path.exists():
                continue
            
            for size_dir in size_dirs:
                icon_dir = hicolor_path / size_dir
                if not icon_dir.exists():
                    continue
                
                for ext in ['.svg', '.png', '.xpm']:
                    icon_path = icon_dir / f"{app_id}{ext}"
                    if icon_path.exists():
                        return str(icon_path)
        
        return None
    
    def _find_in_pixmaps(self, app_id: str) -> Optional[str]:
        """Find in /usr/share/pixmaps"""
        pixmaps = Path("/usr/share/pixmaps")
        if not pixmaps.exists():
            return None
        
        for ext in ['.png', '.svg', '.xpm']:
            icon_path = pixmaps / f"{app_id}{ext}"
            if icon_path.exists():
                return str(icon_path)
        
        return None


def get_resolver() -> IconResolver:
    """Get singleton IconResolver instance"""
    return IconResolver()


def get_nerd_icon(app_id: str) -> str:
    """Convenience function to get Nerd icon"""
    return get_resolver().get_nerd_icon(app_id)


# ═══════════════════════════════════════════════════════════════════════════════
# TESTING
# ═══════════════════════════════════════════════════════════════════════════════

if __name__ == "__main__":
    resolver = get_resolver()
    
    print(f"\n🎨 Current Theme: {resolver._current_theme}")
    
    test_apps = [
        'firefox',
        'kitty',
        'code',
        'spotify',
        'discord',
        'nautilus',
        'steam',
        'unknown-app',
    ]
    
    print("\n📦 Icon Resolution Test:")
    for app in test_apps:
        path = resolver.get_icon_path(app, 24)
        nerd = resolver.get_nerd_icon(app)
        status = "✅" if path else f"⚠️ Nerd: {nerd}"
        print(f"   {app}: {path or 'None'} {status}")